Introduce New Protobuf Definitions for Seat Mapping

havan opened the pull request:
This pull request introduces a new set of Protobuf definitions designed for detailed and flexible seat mapping in systems such as event ticketing or transportation booking.

#### New Features Include:
- **Seat and Feature Representation**: Introduced `Seat` and `SeatAttribute` messages where `SeatAttribute` includes `name`, `description`, and a `value` for dynamic attributes like age restrictions or special conditions. This structured approach allows for more detailed descriptions and easier management of seat features and restrictions.

- **Improved Location Handling**: The `SeatLocation` message supports both SVG and bitmap representations, facilitating precise seat placement on various types of seat maps.

- **Section and Seat Map Structuring**: The `Section` message has been enhanced with recursive capabilities, allowing sections to contain inner sections. This is particularly useful for complex venues with multiple nested areas, such as stadiums or large theaters. `SeatMap` now incorporates the recursive `Section` structure and a localized description set, improving the map's comprehensiveness and usability.

- **Availability & Selection**: Introduced `SeatInventory`, `SectionInventory`, `SeatMapInventory` messages for representing  seat inventory, linked recursively to accommodate complex hierarchical structures of venues (same as the `SeatMap`). These will be used for **seat availability** and **seat selection** messages. Pricing is not included here, it will be in the offers (product search results).

- **SeatMap & Availability Service**: Introduced a simple seat map and seat map availability service in a new package. (Because they will be used by multiple services across different packages like transport and activities)

#### Technical Specifications:
- The schema is designed from the ground up to support both simple and complex seating arrangements across different services.
- The approach ensures that the system can be integrated seamlessly into existing platforms while providing significant flexibility in seat management.
- The `SeatMap` message type is structured to serve as a repository for static data, such as the layout and structural details of seat arrangements. It does not manage dynamic information like seat availability or pricing.
- Dynamic aspects such as availability is managed separately through the inventory message types. This design segregates static and dynamic data, allowing for more efficient updates and queries pertaining to seat status without affecting the underlying seat map structure.

havan committed:
[WIP] Seat map messages

havan committed:
Improve seat map types, add availability and image urls

havan committed:
Update protodot generation script

havan commented:
**Diagram for the new message types**

`types/seat_map.proto`
![seat_map proto dot](https://github.com/chain4travel/camino-messenger-protocol/assets/18965/d9546703-f888-42dd-824b-59d8ac6d8f26)

`seat_map/seat_map.proto` **Seat Map Service**
![seat_map proto dot](https://github.com/chain4travel/camino-messenger-protocol/assets/18965/a6ca7d2a-9c41-49b4-adee-17ee686c84fd)

`seat_map/availability.proto` **Availability Service**
![availability proto dot](https://github.com/chain4travel/camino-messenger-protocol/assets/18965/6973628d-b879-45f2-82c3-0c26f56b585c)

`book/validate.proto` **Validate Service**
![validate proto dot](https://github.com/chain4travel/camino-messenger-protocol/assets/18965/464c4cbb-739a-46e3-9bde-8d582d7594e1)

SamJaarsma commented on proto/cmp/types/v1alpha/seatmap.proto:
```diff
@@ -0,0 +1,171 @@
+
+  // Comma-separated features, e.g., "extra_leg_room,window"
+  // TODO: This field can be improved with a better enum and/or additional fields.
+  string features = 3;
```
features and localized description set feels like a duplication

If we keep features, it should be a repeated string as one specific seat in a plane might have USB charger, window and be close to the toilets. 
A seat in a stadium can be covered or in the open air, designated to the home team supporters. 
A VIP seat in a concert might come with unlimited beverages, merchandise, Meet and greet with the artist
A photo with the artist
Exclusive access to a private pre-show soundcheck
Exclusive access to a backstage tour
Opportunity to shop for artist merch before doors open to the general public
A private performance or Q&A session with the artist
Invitation to a pre- or post-show party
Limited-edition and/or autographed merch
Official VIP laminated pass
On-site VIP concierge team
Free food and drinks
Dedicated VIP entrance and check-in
Of which most will be in the ActivityProductInfo as the package description

havan replied:
Updated the message types and the diagram above

SamJaarsma commented on proto/cmp/types/v1alpha/seatmap.proto:
```diff
@@ -0,0 +1,171 @@
+
+  // Max age for this seat. FIXME: Do we need this?
+  int32 max_age = 5;
+
```
min age and max age are not the only conditions, so this should be a repeated string. For example for an emergency exit seat from southwest airlines:

Be 15 years of age or older
Have the capacity to perform the applicable functions without the assistance of an adult companion, parent, or other relative
Read and understand instructions related to emergency evacuation provided by Southwest in printed or graphic form
Understand oral crew commands
Have sufficient visual capacity to perform applicable functions without the assistance of visual aids beyond contact lenses or eyeglasses
Be able to hear and understand instructions shouted by Flight Attendants without assistance beyond a hearing aid
Adequately impart information orally in English to other passengers

havan replied:
Updated the message types and the diagram above

havan committed:
Use recursive structure for seat grouping

havan committed:
remove boolean for image location from section

havan committed:
add missing map_id to the availability message

havan committed:
split features and attributes

havan committed:
rename availability messages

havan committed:
add seatmap service to transport package

havan committed:
fix enum truncation for multi enum files for diagrams

havan committed:
use correct naming pattern for fields and filenames

havan committed:
move seat map service and add availability service

SamJaarsma commented on proto/cmp/services/seat_map/v1alpha/seat_map.proto:
```diff
@@ -0,0 +1,49 @@
+package cmp.services.seat_map.v1alpha;
+
+import "cmp/types/v1alpha/common.proto";
+import "cmp/types/v1alpha/seat_map.proto";
```
this import of seat_map.proto into seat_map.proto is obsolete, correct?

havan replied:
No, this is from the `seat_map.proto` from types package. And this file is in its own package, which also called `seat_map`.

BTW, the redundant or missing imports are checked by the linter.

SamJaarsma commented on proto/cmp/services/seat_map/v1alpha/seat_map.proto:
```diff
@@ -0,0 +1,49 @@
+
+  // Unique identifier for the seat map
+  //
+  // This is the map ID that is received in the search result
```
// This is the map ID that is received in the Search result and in the ProductInfo response

havan replied:
This comment refers to the the `map_id` that is returned in the response below (in the same file). Why should we add the ProductInfo there? Not all product info messages will have a seat map id.

SamJaarsma replied:
not all seatmap ids will have a ProductInfo either. But if we want to describe anything about a section or a seat, the link is the map id path, right?

havan replied:
Ok, I thought you were saying to add a field for product info. 😅  I've updated the comment.

SamJaarsma commented on proto/cmp/types/v1alpha/seat_map.proto:
```diff
@@ -0,0 +1,147 @@
+package cmp.types.v1alpha;
+
+import "cmp/types/v1alpha/description.proto";
+import "cmp/types/v1alpha/price.proto";
```
price is not required

havan replied:
Updated

SamJaarsma commented on proto/cmp/types/v1alpha/seat_map.proto:
I still find it confusing to have the same filename in proto/cmp/types and in proto/cmp/services/seatmap

havan replied:
It's in another package. What's your suggestion?

havan committed:
remove price from seat availability

havan committed:
add integer availability info to section

havan committed:
update validate message to include seat info
refactor availability message names to inventory
update availability service with different identifiers

havan committed:
remove SeatType message

havan committed:
update inventory system for section and seat
- SeatInventory updated to be a a list of seat_ids
- SectionInventory now contains a oneof field of seats that has seat_inventory and seat_counts fields.
- SeatCounts msg type contains total and remaining seat counts

havan committed:
update validation messages
- validation messages now uses the validation object and search identifier
- this removes redundant code and makes it simpler

havan committed:
update comments for the seat map service

Noctunus commented on proto/cmp/types/v1alpha/seat_map.proto:
```diff
@@ -0,0 +1,154 @@
+message Section {
+  // Level, section or row identifier, e.g., "Upper", "Balcony" or "Section 101",
+  // "Orchestra" or "A", "12".
+  string id = 1;
```
For maps it's `map_id`, for seat it's `seat_id`, for section it's `id`?
We should have it consistent. I'm in favor of getting rid of the prefix in case it's inside of the message the id is for (which is the same for other fields like name).

havan replied:
I don't know how that one slipped, I was using `<name>_id` for all. But yes, I'm also in favor of loosing the prefix.

Noctunus commented on proto/cmp/types/v1alpha/seat_map.proto:
```diff
@@ -0,0 +1,154 @@
+// Total and remaining number of seats in a section.
+message SeatCounts {
+  // Total number of seats
+  int32 total_seat_count = 1;
```
We can get rid of the `_seat_count` suffix as it's clear from the message it's in what "total" and "remaining" refers to. What do you think?

havan replied:
I agree!

havan replied:
Updated. I introduced a new structure with `oneof` fields and separating  the static (total) and dynamic (remaining) fields into corresponding messages.

havan committed:
loose the prefix and update comments

havan committed:
move seats and total/remaining seats into oneof fields
- comments updated/improved
- SeatMapLocation renamed to SeatLocation
- image_urls replaced with the Image type from file.proto
- remaining_seats renamed to seat_count for availability & seat selection
- helper types moved to the end of the file

havan committed:
update diagram script and add wrappers.proto include

Noctunus reviewed (approved):
LGTM
